pkg/flow/streaming/flow_sliding_window.go

```diff
@@ -0,0 +1,334 @@
+			}
+			s.snapshotsMu.Unlock()
+
+			result := ctx.OnElement(elem)
+			if result == FIRE {
+				s.purgedWindow <- tw
+			}
+		}
+
+		// even if the incoming elements do not follow strict order,
+		// the watermark could increase monotonically.
+		if elem.TimestampMillis() > s.currentWatermark {
+			s.currentWatermark = elem.TimestampMillis()
+
+			// TODO: add various strategies to allow lateness items, which come later than the current watermark
+			// Currently, assume the current watermark is t,
+			// then we allow lateness items coming after t-windowSize+slideSize, but before t,
+			// i.e. [t-windowSize+slideSize, t)
+			// NOTE: This call is the normal path to flush outdated windows immediately
+			// without considering lateness.
+			// The purged windows may be accumulated again due to non-zero lateness.
+			s.purgeOutdatedWindows()
+		}
+	}
+	close(s.purgedWindow)
+	<-s.done
+	close(s.out)
+}
+
+// isWindowLate checks whether this window is valid
+func (s *TumblingTimeWindows) isWindowLate(w flow.Window) bool {
+	return w.MaxTimestamp()+s.lateness <= s.currentWatermark
+}
+
+func (s *TumblingTimeWindows) Teardown(ctx context.Context) error {
+	s.Wait()
+	return nil
+}
+
+func (s *TumblingTimeWindows) Exec(downstream flow.Inlet) {
+	s.Add(1)
+	go flow.Transmit(&s.ComponentState, downstream, s)
+}
+
+func NewTumblingTimeWindows(size time.Duration) *TumblingTimeWindows {
+	return &TumblingTimeWindows{
+		size:             size.Milliseconds(),
+		lateness:         0,
+		snapshots:        make(map[int64]flow.AggregationOp),
+		timerHeap:        flow.NewPriorityQueue(false),
+		in:               make(chan flow.StreamRecord),
+		out:              make(chan flow.StreamRecord),
+		done:             make(chan struct{}),
+		purgedWindow:     make(chan timeWindow),
+		currentWatermark: 0,
+	}
+}
+
+type timeWindow struct {
+	start int64
+	end   int64
+}
+
+func (t timeWindow) MaxTimestamp() int64 {
+	return t.end - 1
+}
+
+// AssignWindows assigns windows according to the given timestamp
+func (s *TumblingTimeWindows) AssignWindows(timestamp int64) ([]flow.Window, error) {
+	if timestamp > math.MinInt64 {
+		start := getWindowStart(timestamp, s.size)
+		return []flow.Window{
+			timeWindow{
+				start: start,
+				end:   start + s.size,
+			},
+		}, nil
+	}
+	return nil, errors.New("invalid timestamp from the element")
+}
+
+// getWindowStart calculates the window start for a timestamp.
+func getWindowStart(timestamp, windowSize int64) int64 {
+	remainder := timestamp % windowSize
+	return timestamp - remainder
+}
+
+// eventTimeTriggerOnElement processes element(s) with EventTimeTrigger
+func eventTimeTriggerOnElement(window timeWindow, ctx TriggerContext) TriggerResult {
+	if window.MaxTimestamp() <= ctx.GetCurrentWatermark() {
+		// if the watermark is already past the window fire immediately
+		return FIRE
+	}
+	ctx.RegisterEventTimeTimer(window.MaxTimestamp())
+	return CONTINUE
+}
+
+type TriggerContext interface {
```

Comment:
What's this exported interface used for? Its invokers are from this package. I don't find other implements more than triggerContext.

Reply:
Removed redundent interface